src/java.base/share/classes/java/util/zip/ZipFile.java

```diff
@@ -1734,12 +1736,12 @@ private void initCEN(int knownTotal) throws IOException {
                 if (locpos < 0) {
                     zerror("invalid END header (bad central directory offset)");
                 }
-                // read in the CEN and END
-                if (end.cenlen + ENDHDR >= Integer.MAX_VALUE) {
+                // read in the CEN
+                if (end.cenlen > MAX_CEN_SIZE) {
```

Comment:
We'd be throwing an OOMError today if we soared too close to the limit (`Integer.MAX_VALUE - ENDHDR - 2` and above), then throw `zerror` if we go beyond the limit. With these changes we change to throw `zerror` exclusively. An OOME can of course still be thrown from any subsequent allocation on a _real_ OOM, so I'm not sure how valuable this edge case trimming is.

Reply:
> We'd be throwing an OOMError today if we soared too close to the limit (Integer.MAX_VALUE - ENDHDR - 2 and above), then throw zerror if we go beyond the limit. 

I think you might be off by one there :)

The current code incorrectly fails to reject a CEN size of `Integer.MAX_VALUE - ENDHDR - 1`, and instead cause an OOME for exceeding the implementation limit of int[]. The size `Integer.MAX_VALUE - ENDHDR - 2` is allowed, as it should.

So the existing code was off-by-one in that it allowed exactly one CEN size which was larger than the implementation limit.

Yes, I'd be surprised if this fix gets any user excited. But I think from a maintainer's point of view, it may be easier to understand code when it's stricly correct since it removes any reason to ponder why the code is almost, but not completely correct.